src/core/ObservableQuery.ts

```diff
@@ -1174,121 +1366,59 @@ Did you mean to call refetch(variables) instead of refetch({ variables })?`,
-        oldNetworkStatus !== newNetworkStatus &&
-        isNetworkRequestInFlight(newNetworkStatus),
       query
     );
-    const observer: Partial<Observer<ApolloQueryResult<TData>>> = {
-      next: (result) => {
-        if (equal(this.variables, variables)) {
-          finishWaitingForOwnResult();
-          this.reportResult(result, variables);
-        }
-      },
-      error: (error) => {
-        if (equal(this.variables, variables)) {
-          finishWaitingForOwnResult();
-          this.reportError(error, variables);
-        }
-      },
-    };
 
     if (!useDisposableObservable && (fromLink || !this.linkSubscription)) {
       if (this.linkSubscription) {
         this.linkSubscription.unsubscribe();
       }
 
-      this.linkObservable = observable;
-      this.linkSubscription = observable.subscribe(observer);
-    } else {
-      observable.subscribe(observer);
+      this.linkSubscription = subscription;
     }
 
     return preventUnhandledRejection(
       // Note: lastValueFrom will create a separate subscription to the
       // observable which means that terminating this ObservableQuery will not
       // cancel the request from the link chain.
-      lastValueFrom(observable, {
-        // This default value should only be used when using a `fetchPolicy` of
-        // `standby` since that fetch policy completes without emitting a
-        // result. Since we are converting this to a QueryResult type, we
-        // omit the extra fields from ApolloQueryResult in the default value.
-        defaultValue: { data: undefined } as ApolloQueryResult<TData>,
-      }).then((result) => toQueryResult(this.maskResult(result)))
-    );
-  }
-
-  // (Re)deliver the current result to this.observers without applying fetch
-  // policies or making network requests.
-  private observe() {
-    this.reportResult(
-      // Passing false is important so that this.getCurrentResult doesn't
-      // save the fetchMore result as this.lastResult, causing it to be
-      // ignored due to the this.isDifferentFromLastResult check in
-      // this.reportResult.
-      this.getCurrentFullResult(false),
-      this.variables
+      lastValueFrom(
+        observable.pipe(
+          filterMap((value) => {
+            switch (value.kind) {
+              case "E":
+                throw value.error;
+              case "N":
+                if (value.source !== "newNetworkStatus") return value.value;
+            }
+          })
+        ),
+        {
+          // This default value should only be used when using a `fetchPolicy` of
+          // `standby` since that fetch policy completes without emitting a
+          // result. Since we are converting this to a QueryResult type, we
+          // omit the extra fields from ApolloQueryResult in the default value.
+          defaultValue: { data: undefined } as ApolloQueryResult<TData>,
+        }
+      ).then((result) => toQueryResult(this.maskResult(result)))
```

Comment:
```suggestion
          filterMap((value) => {
            switch (value.kind) {
              case "E":
                throw value.error;
              case "N":
                if (value.source !== "newNetworkStatus") {
                  return toQueryResult(this.maskResult(value.value));
                }
            }
          })
        ),
        {
          // This default value should only be used when using a `fetchPolicy` of
          // `standby` since that fetch policy completes without emitting a
          // result.
          defaultValue: { data: undefined } as QueryResult<TData>,
        }
      )
```

Now that the code is structured where you have access to the `ApolloQueryResult` inside the `pipe` above, it might be worth removing this `.then` and converting the result to a query/masked result in the `filterMap` directly. That would allow us to properly cast that `defaultValue` as a `QueryResult` as well.

Reply:
This was actually deliberate - we are inside a `lastValueFrom`, so if we move this in and there are 5 values emitted, `this.maskResult` will be called 5 times but we only care about the last one.
This way it will only do the calculations on the last emitted value, which should help a bit with performance.

Reply:
Fair. Let's leave it then.